### PR TITLE
Implement subscription API functions

### PR DESCRIPTION
**Pull request by pavel-kirienko**
It is now possible to subscribe to subjects and receive data from them. The only missing bit is the RPC services API, all other functionality of the library is already implemented and can be used.

This changeset also brings `udpardGather` that, well, gathers a fragmented payload buffer into one contiguous chunk of memory that can be passed into Nunavut. Later we should amend Nunavut to support fragmented buffers directly to eliminate extra data copying and memory utilization.

After the RPC services API is done, I am going to add comprehensive end-to-end tests, for which there is a stub already. Aside from that we also have this (the e2e test mentioned there is different though): https://github.com/OpenCyphal-Garage/libudpard/issues/35

This changeset also fixes an internal documentation error that provided an incorrect estimation of the worst-case recursion depth.

### CODE REVIEW

File: libudpard/udpard.h [hunk 1]
```diff
@@ -1021,6 +1020,22 @@ int_fast8_t udpardRxRPCDispatcherReceive(struct UdpardRxRPCDispatcher* const sel
...
 
+// =====================================================================================================================
+// ====================================================    MISC    =====================================================
+// =====================================================================================================================
+
+/// This helper function takes the head of a fragmented buffer list and copies the data into the contiguous buffer
+/// provided by the user. If the total size of all fragments combined exceeds the size of the user-provided buffer,
+/// copying will stop early after the buffer is filled, thus truncating the fragmented data short.
+///
+/// The source list is not modified. Do not forget to free its memory afterward if it was dynamically allocated.
```
Comment (thirtytwobits) [inline libudpard/udpard.h:1031]:
```
/// also remember to change the air filter in your AC system every six months 
/// and the batteries in your smoke detectors every year.
```

File: tests/src/test_misc.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    frags.at(2).view.data = payload.data() + frags.at(1).view.size + frags.at(0).view.size;
+    frags.at(2).view.size = 0;  // Edge case.
+
+    frags.at(3).view.data = payload.data() + frags.at(2).view.size + frags.at(1).view.size + frags.at(0).view.size;
+    frags.at(3).view.size = payload.size() - frags.at(2).view.size - frags.at(1).view.size - frags.at(0).view.size;
+
+    std::array<std::uint8_t, 1024> mono{};
+
+    // Copy full size payload.
+    std::generate(mono.begin(), mono.end(), [] { return std::rand() % 256; });
```
Comment (thirtytwobits) [inline tests/src/test_misc.cpp:45]: To generate random test bits use `std::independent_bits_engine` and be sure to provide a stable seed so the tests are reproducible. I tend to use googletest, value-parameterized for this where I run the test using a few seeds to really get the psedo-random mojo going but I'm fine with just using unity here and a single seed value.

File: libudpard/udpard.c [hunk 3]
```diff
@@ -1642,12 +1654,67 @@ int_fast8_t udpardRxSubscriptionInit(struct UdpardRxSubscription* const   self,
...
+    size_t offset = 0;
+    if (NULL != destination)
+    {
+        const struct UdpardFragment* frag = &head;
+        while ((frag != NULL) && (offset < destination_size))
+        {
+            UDPARD_ASSERT(frag->view.data != NULL);
+            const size_t frag_size = smaller(frag->view.size, destination_size - offset);
+            // NOLINTNEXTLINE(clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling)
+            (void) memmove(((byte_t*) destination) + offset, frag->view.data, frag_size);
```
Comment (thirtytwobits) [inline libudpard/udpard.c:1713]: Do we define byte_t as unsigned char? I forgot if this was a local typedef or not.

File: libudpard/udpard.c [hunk 4]
```diff
@@ -1642,12 +1654,67 @@ int_fast8_t udpardRxSubscriptionInit(struct UdpardRxSubscription* const   self,
...
+                                   out_transfer);
+    }
+    return result;
+}
+
+// =====================================================================================================================
+// ====================================================    MISC    =====================================================
+// =====================================================================================================================
+
+size_t udpardGather(const struct UdpardFragment head, const size_t destination_size, void* const destination)
```
Comment (thirtytwobits) [inline libudpard/udpard.c:1702]: `%s/destination_size/destination_size_bytes/`

**Review by thirtytwobits — APPROVED** [review]

File: libudpard/udpard.h (hunk 1 above)
Comment (pavel-kirienko) [reply libudpard/udpard.h:1031]: I sometimes wonder if the concept of the fortune cookie is applicable to documentation comments.

File: libudpard/udpard.c (hunk 3 above)
Comment (pavel-kirienko) [reply libudpard/udpard.c:1713]: It is internal:

https://github.com/OpenCyphal-Garage/libudpard/blob/c92f8df4c3be2a3c0e6df501d97576a2df8e8080/libudpard/udpard.c#L33

File: tests/src/test_misc.cpp (hunk 2 above)
Comment (pavel-kirienko) [reply tests/src/test_misc.cpp:46]: Indeed, I overlooked the seeding part. However, I think it would be unwise to use the C++ RNG utilities here because this is a C library and most of the test code is written in C, so in the interest of consistency and compatibility with future tests written in C, it is desirable to use `rand()`. The quality of the random numbers matters little in this case.

I implemented seeding in the last commit.

**Review by lydia-at-amazon — COMMENTED** [review]
Sorry for the very late review, looks good! Excited to see the final API implemented and glad we added the `udpardGather` function. Can't approve anymore, can only give you two thumbs up: 👍 👍

File: tests/src/test_intrusive_rx.c [hunk 5]
```diff
@@ -332,6 +332,15 @@ static void testParseFrameEmpty(void)
     TEST_ASSERT_FALSE(rxParseFrame((struct UdpardMutablePayload){.data = "", .size = 0}, &rxf));
 }
 
+static void testParseFrameInvalidTransferID(void)
+{
+    byte_t  data[] = {1,   2,   41,  9,   56, 21, 230, 29, 255, 255, 255, 255,
+                      255, 255, 255, 255, 57, 48, 0,   0,  0,   0,   42,  107,  //
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply tests/src/test_intrusive_rx.c:338]: No, I sometimes add these to hint Clang-Format how I want the code to be formatted. A comment like this forces a line break.
